### DISCUSSION TITLE
GPIO on NXP MIMXRT1040-EVK

### DISCUSSION THREAD

**Discussion by DerekSnell** [zephyrproject-rtos/zephyr#56777] [Q&A] [answered]
Creating a discussion started by @matt-wood-ct at https://github.com/zephyrproject-rtos/hal_nxp/pull/229#issuecomment-1496620732:
> Additionally we found that the DR_SET/TOGGLE/CLEAR register appear not to work on the current die revision, is NXP aware of this? We attached to the EVK using a Segger and found that even manually writing to those registers had no effect on the actual GPIO. In the end we unasserted that feature in the features.h

I am not aware of this issue.  Is there a reason you feel this is related to the current die revision?  The device on my MIMXRT1040-EVK board is revision 1N00X (printed on the package).  And I see the DR_SET/TOGGLE/CLEAR registers controlling the GPIO as expected.

--edited after finding the cause of the GPIO conflict--
I was able to test these registers using [Zephyr's blinky sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/basic/blinky).  NXP submitted a [Pull Request](https://github.com/zephyrproject-rtos/zephyr/pull/57358) to add board support for the MIMXRT1040-EVK to the Zephyr repo.  In the meantime, this EVK and SOC is very similar to the MIMXRT1060-EVKB board.  So I built  for the board `mimxrt1060_evkb` and ran it on the MIMXRT1040-EVK board.  This app toggles the USER_LED D8 as expected.  Note that the MIMXRT1040-EVK ships with a wireless module in the M.2 connector J8.  The GPIO to USER_LED is shared with this module.  I had to remove the module or remove jumper J80 to see USER_LED toggle.

When the app toggles the LED, the HAL API `GPIO_PortToggle` [writes to the DR_TOGGLE register](https://github.com/zephyrproject-rtos/hal_nxp/blob/dd6016b073f2f097e157f4ce26ecd36792ff339d/mcux/mcux-sdk/drivers/igpio/fsl_gpio.h#L163).  So this demonstrates the DR_TOGGLE register works with the RT1040 1N00X revision.

How does this compare to what you have found?  Thanks

**Comment by matt-wood-ct** [comment]
Hi Derek,
Thanks for your reply,

RE the DR_xx behaviour I only suspected the die revision because it may be a chip issue as opposed to a driver issue, but I am unsure what is going on with that behaviour.
I did have luck with the 1060 EVK board on the 1040 DK but again I did need to remove the feature flags for the DR_xx functions. On our DKs we did find that J80 was fitted out of the box which appears to be a mistake because the EVK schematic says its DNF and in practice having it fitted prevented proper operation of the LED as the line is connected through to the wifi module. This may be why you are seeing issues with the LED on the EVK.

Best Regards,
Matt

**Comment by DerekSnell** [comment]
Hi @matt-wood-ct ,

Thank you for the response.  You are right, the wireless module and J80 were the cause of the LED issue I had.  I have edited my post above to clean that up.

I am still unclear why you have issues using these DR_ registers.  I do not see the problem.  I also downloaded the latest v2.13.1  [MCUXpresso SDK package for the MIMXRT1040-EVK board](https://mcuxpresso.nxp.com/en/builder?hw=MIMXRT1040-EVK), and tested the example \boards\evkmimxrt1040\demo_apps\led_blinky.  This app uses the DR_SET and DR_CLEAR registers, and they work well for me.  In addition to the LED toggling as expected, below are screenshots of the debugger watching the GPIO hardware registers.  I see the the corresponding bit in the DR register clearing when DR_CLEAR is written.

Are you able to provide instructions for me to replicate the issue you see using our MIMXRT1040-EVK board?  Thanks

Before writing to DR_CLEAR:
![image](https://user-images.githubusercontent.com/78033768/231507569-86e4a91f-29e5-4051-bae0-53c90a8cad7b.png)

After writing to DR_CLEAR:
![image](https://user-images.githubusercontent.com/78033768/231506324-92c916cf-9202-4ce4-b4ee-3d68b057679f.png)

**Reply by matt-wood-ct** [reply, marked as answer]
Hmm, thanks for testing this. 
I just re-enabled those feature flags again and the the IO are working (using my HAL_NXP fork). 
I'm not sure why I found DR_SET to not work previously, perhaps there was something else wrong with my setup at the time.
I did reattach with a segger and tried writing 0xffffffff into the DR_TOGGLE for GPIO 1 and I did note the LED flashing each time I wrote to the register.
All is well 😄 

BR,
Matt
